### DISCUSSION TITLE
Normal filtering in TIDDIT

### DISCUSSION THREAD

**Discussion by mathiasbio** [Clinical-Genomics/BALSAMIC#1127] [General]
## Background 

In issue https://github.com/Clinical-Genomics/BALSAMIC/issues/1118 it was brought up that there has been a significant increase in the number of SV calls in the final VCF since version 10 when TIDDIT was introduced, and which is due primarily to the lack of filtering of normal variants in the merged tumor + normal VCF.

This is attempted to be solved in this PR: https://github.com/Clinical-Genomics/BALSAMIC/pull/1120 by implementation of filters in bcftools which filters variants based on their presence in the normal.

But before implementing the filtering I wanted to report the results of this filtering, and allow for feedback based on this. 

This filtering is only relevant for T / N WGS cases, and I selected 3 at random to look at: A, B and C, all with about 13k variants each in the merged vcf from TIDDIT without any filters applied. (See original caseIDs in private CG google drive doc: https://docs.google.com/document/d/1dP1mM-LkuvZUOA6cripXC5I089V-M8knEw26W5hBMEs/edit)

## Results 

Out of the total 40117 variants looked at, 2 had variant-frequencies that gave me some concern for the viability of calculating allele-frequencies from TIDDIT, as they had allele frequencies in the normal above 1, which I would not expect to be possible.

_For downstream plotting and reporting these 2 variants are excluded._

Below is a scatterplot of all variants in all 3 samples, based on allele frequency in normal (Y) and tumor (X).

![case_af_t_af_n](https://user-images.githubusercontent.com/14308912/229897607-5a7c7b4e-c91a-4ef4-8fc3-dcc9512c0a17.png)

After applying filters in the PR: 
1. Mark all variants which have been called as PASS in the normal with "in_normal"
2. All variants with AF_T >= 0.4 & AF_N <= 0.1 is set as "tumor_contam..." and will be rescued --> PASS.
 
![v_filter_af_t_af_n](https://user-images.githubusercontent.com/14308912/229750597-1220df38-4398-4eb5-8764-df4f275101ac.png)

Reasoning for the filter set in this way:
I had preferred to implement a filter where the allowed frequency in the normal was based on a maximum tumor contamination level of 0.1. Such that:
- tumor AF = 0.9, allowed normal AF <= 0.09
- tumor AF = 0.5, allowed normal AF <= 0.05
- tumor AF = 0.2, allowed normal AF <= 0.02
- etc...
Like this:
![v_filter_af_t_af_n_clean](https://user-images.githubusercontent.com/14308912/229760694-fd76c1d6-7749-4191-bba7-f178313ce704.png)

However it doesn't seem like bcftools is capable of implementing a filter which depends on multiple samples within the same vcf in this way, and I was forced to rely on a fixed value. So I chose a space of allele-frequencies that would allow for tumor contamination between 25-10%. 

![v_filter_af_t_af_n](https://user-images.githubusercontent.com/14308912/229760407-2c78993d-c565-4df9-a4c0-3110b1eac31e.png)

There are 4 groups of variants in this merged vcf based on AF_T and AF_N, and I interpret them like this:
1. Primarily true somatic variants present in the normal, acquired through life but not present in the tumor.
2. A mix of true germline variants present in both the tumor and the normal and artifacts due to problematic regions in the genome
3. Uncertain variants, mix of true low frequency somatic variants, and artifacts
4. Primarily true somatic variants present in the tumor, and not present in the normal except as potential contamination from the tumor

![v_filter_af_t_af_n_clean2](https://user-images.githubusercontent.com/14308912/229769122-d0dda761-5802-4fc7-a06e-94901776bc44.png)

### Contig variants

It's a bit difficult to see in these plots the overlapping variants, and there are some interesting ones left to mention...

While it appears from the plots above that all variants with AF_N = 0 has PASS, there are variants with the filter "in_normal" that have an allele frequency in the normal sample of 0. Yet they have been called in the normal. 

In fact out of 40115 variants, 3311 of them fit that description. The same is true for the tumor variants, where 2499 of the variants have an AF_T of 0 and yet are still called as PASS.

This seems to have to do with how TIDDIT reports certain variants that have successfully been assembled into a contig, where the stats are sometimes lost.

In the scatterplot below I have included only variants with filter "in_normal", and the colors marks if they have a contig in the normal or not.

![CTG_N_in_normal_af_t_af_n_0 05](https://user-images.githubusercontent.com/14308912/229770953-97931b0b-c85a-4170-8afb-cbf85f2796a3.png)

In addition I can confirm that:
- All variants flagged as "in_normal" with an AF_N of 0 has a contig assigned in the normal
- All variants with "PASS" and an AF_N of 0 doesn't have a contig assigned in the normal.

I discussed these variants with Jesper, the creator of TIDDIT, last week and he claimed that it was unlikely that a variant could have an assembled contig without significant read support, and based on that it would be safe to remove them as normal variants.

Some of these variants are still rescued in the contamination filtering, and in the scatterplot below only variants with a contig in the normal are shown, the orange marked variants will be rescued: 
![v_filter_af_t_af_n_CTGN](https://user-images.githubusercontent.com/14308912/229779949-d2791524-6d46-4e7b-a7c0-fdbf4af77fc8.png)

## Discussion

Is there a significant risk that I'm filtering away true somatic variants that should be taken into account here? 

Risk factors identified:
1. The tumor contamination may be higher than 10%
2. Variants below 0.4 AF_T allow no presence of the variant in the normal. This should be safe in most cases as the normal is sequenced at an average of 30x and have a minimum of 3 supporting reads required to call a variant. Let's assume a variant with 0.39 AF_T, in normal with 10% contamination AF_N 0.039. Sequenced at an average of 30x, only 30 * 0.039 = 1,17 reads support the variant and it should not be called. By chance the minimum support of 3 reads may be achieved and be called in the normal and be filtered out. 
3. The presence of the variant in the normal where contigs have been assembled is unknown, perhaps for a high frequency variant there may have been enough contamination to create a contig, but at a low enough frequency to not be otherwise filtered out. Example...a variant present at 90% in the tumor, expecting 10% contamination in normal, AF_N would be 0.09. Region with unusual high coverage for a normal, 60x: 60 * 0.09 = 5,4 reads. If that was enough to create the contig it could filter out the variant despite being present below 10%...but this is a lot of IFs...

In general I think this is a safe filter to implement to reduce the amount of noise. At least in these cases it does not appear to be a lot of tumor in the normal contamination (TINC), if there was a significant TINC I would expect to see less of a clear separation between the shared and unique variants.

**Comment by pbiology** [comment]
Very nice write-up! 

A question about the 4 different kind of variants.
"1. Primarily true somatic variants present in the normal, acquired through life but not present in the tumor."

Is the normal always from buffy coats? In that case, would one expect to see less of this kind where the tumor comes from a blood samples vs say a solid tumor? What are the types of tumors in these 3 cases?

**Reply by mathiasbio** [reply]
I don't know if I'm ignorant or buffy coats = blood sample! I'll give myself some credit and assume it means blood sample. 

It is generally known that haemotlogical cancers have a greater tumor contamination than solid tumors at least, and in that case I would expect that variants in group 1 and 4 would be more pushed towards the center. 

I tried to read up on what the tumor in normal contamination fraction usually is earlier today, and from what I've seen so far it seems quite variable. But that at least around 38% of saliva samples from AML  and in MPN 91% of sorted CD3-T cells had Tumor In Normal (TIN) values above 1%.  For sarcoma samples only 2% of samples were contaminated with TIN above 2%. https://www.biorxiv.org/content/10.1101/2022.03.09.483623v1.full.pdf 

But I found it difficult to find what the actual fraction commonly is.

I don't actually know exactly but at least:
- B is uploaded to "Akuta leukemier" (normal from skin, tumor from bonemarrow)
- A OnkPat BTB/GMS (normal from blood, tumor from tissue fresh frozen)
- C is uploaded to "Akuta leukemier" (normal from saliva, tumor from bonemarrow)

Perhaps the people taking these samples are aware of the contamination problem and is avoiding taking the normal from blood in the case of haematological cancers

**Reply by vwirta** [reply]
Buffy coat is derived from blood and contains the white blood cells and platelets. 

Regarding contamination of Normal sample with Tumour... it is possible that this happens, and it is good that we consider this. 

**Skin:** The reason for using skin in hematological malignancies is to reduce the quantity of tumor in normal. A skin biopsy per se can still contain some blood, hence a small risk still remains. If the lab has cultured fibroblasts from skin biopsies, then there are no more tumour cells. This would be the safest normal to use, but requires both an invasive skin punch + 4-6 days of culturing. 

**Saliva:** Does contain tumor DNA and at variable frequency. In many cases low one-digit percentage numbers. >10% is rare afaik. We can bring this up with clinical experts for further discussion.

**Blood/Buffy coat for solid tumours:** Not much contamination expected, but can never be entirely ruled out.

**Comment by mathiasbio** [comment]
**Here's an update on this filtering.**

Issues regarding the filtering:
* Variants with assembled contigs have unreliable AF values, sometimes at 0 (and in 2 cases above 1). How does this affect the tumor / normal filtering?
* BND variants are present once per breakpoint, 1 of them may be overlapping with a normal BND, the other not. Bcftools can only filter with data present in the variant row. How do we avoid filtering away only one BND? (both are needed for VEP)
* Are the tumor / normal filtering properly set given we want to rescue potential tumor in normal contamination of 10%?

## Filtering based on AF_T and AF_N, accounting for 10% TIN

The earlier mentioned filters was fine for the 3 cases investigated, and reported earlier. But these cases were not good examples of tumor contamination. 

Purely based on logic it was possible to see that the allowing for a 10% of the AF_T in the normal was a bit too stringent if the goal was indeed to allow for 10% TIN, as while the average AF_N of these variants would be AF_T * 0.1, there should of course be a spread based on randomness.

I did some simulations of the presence of tumor variants in the normal data to have a better idea of what we might expect, and at what thresholds to put the filtering. 

Given preset list of AF_T, sequencing depth in the normal, and a TIN of 10% a scatterplot of the AF_N and AF_T can be seen below:

<img width="744" alt="image" src="https://user-images.githubusercontent.com/14308912/231135989-68d2f453-b0b7-4709-b84b-88e37231b980.png">

For 40% TIN the presence of the AF in the normal is predictably much greater. 

<img width="752" alt="image" src="https://user-images.githubusercontent.com/14308912/231136553-1981fa1c-cba3-47c5-a470-19cf06d9cea2.png">

With a filter of  AF_N / AF_T > 0.25 = IN_NORMAL, and thereby allowing for a maximum of 25% TIN, and accounting for the spread. The result on this simulated data can be seen below for a 10% TIN:

<img width="718" alt="image" src="https://user-images.githubusercontent.com/14308912/231137339-01de8c32-702f-497b-a960-7bb4c85e9852.png">

Which comes down to: 1.19% filtered out somatic variants

However, the average coverage of these variants does not come to 30x normal, but 89.42x. And if we limit the coverage to exactly 30x in the normal the results are a bit worse, but still ok in my mind:

<img width="662" alt="image" src="https://user-images.githubusercontent.com/14308912/231139944-a5ec27af-0e3c-44e5-b526-914da12ac859.png">

With around 2.25% falsely filtered out variants at a TIN of 10%.

What does applying this filter look like in the 3 cases (A, B, C)?

<img width="1223" alt="image" src="https://user-images.githubusercontent.com/14308912/231140380-cc173f0f-859b-440d-b014-917dbe7caaed.png">

And the diagonal line of the filter is somewhat outlined by the separation of the "PASS" and "IN_NORMAL" variants, and you can imagine it going from 0.0 diagonally to 0.25 of AF_N. 

This looks fine to me as it avoids the mess of variants in the center which should mostly be germline and artifacts, while still allowing for an average of 10% TIN. 

The average coverage in the normal of the simulated variants filtered out as "in_normal" was a bit lower than those with "PASS", 55.17x in comparison to 89.83x. This trend is expected given that the spread of AF_N should be higher the lower you sequence, and would increase the chance that a TIN of 10% reached a higher level than 25%, which is the maximum allowed here.

Below is a scatterplot of the "in_normal" variants, with coverage in the normal on the X-axis. 

<img width="734" alt="image" src="https://user-images.githubusercontent.com/14308912/231147956-a54c6d20-04f6-424c-bd97-6b550e7be0fb.png">

Perhaps some additional filter could be implemented to rescue these lower coverage variants as well. It should be noted again however that even without this additional rescue filter, the number of falsely filtered out variants at 10% TIN is 2.25%, and a TIN of 10% should be pretty rare.

## Contigs

I brought this up earlier as well...

The existence of these variants with assembled contigs make the filtration a little bit more complex as the AF values don't always make sense.

For instance, there are in these 3 cases:
* 2985 / 40115 variants that have a contig in the tumor, and an AF_T of 0.
* 3311 / 40115 variants that have a contig in the normal, and an AF_N of 0.
Beyond that are those 2 previously mentioned variants with AF_N above 1, which should be impossible. This makes it a bit less straight-forward to filter variants while allowing for a possible TIN, as it is a bit uncertain how much support the variant truly has in the tumor and the normal.

For instance, if 3 reads in the normal is sufficient to construct the contig, and the position has been sequenced at 30x, it would have an AF_N of 0.1. If the corresponding variant in the tumor has a frequency of 90% this would be within the allowed TIN of 25%. But we don't have any data on this. 

In the above example, the AF_N and the AF_T may both be 0, and the consequence of this uncertainty of AFs is that the allowed presence in the normal of variants with and without the assembled contigs **may** be different. 

I mentioned this earlier too, and referenced a discussion with Jesper where the conclusion was that a variant with a contig in the normal was most likely a normal variant. Speaking to him again, it was possible according to him that as few as 3 reads may be sufficient in some cases to construct the contig. 

In a scenario of 10% TIN, this would put quite a number of variants with contigs in the normal at risk of being filtered out. 

The question is; is it worth accounting for these possibilities and not filter variants with contigs at all? 

To start with, we can at least filter out all variants without either a contig or any AF in the tumor, beyond that we can set to PASS all variants with neither any contig or AF in the normal.

<img width="1246" alt="image" src="https://user-images.githubusercontent.com/14308912/231191535-20275389-e953-43e8-9f6e-9d79eee13ae2.png">

After applying these most loose of filters there's a significant number of "Uncertain" variants with a presence in the normal and the tumor which we may want to filter.

How many of these have a contig?  

<img width="1248" alt="image" src="https://user-images.githubusercontent.com/14308912/231193099-0c5ca388-5180-4904-b137-a1dcc6185352.png">

The variants in the blue box have no contig in either tumor or normal, and can be filtered simply with the AF filters previously discussed, and then we have the remaining variants.
* Red: With contigs in both
* Green: with contig in tumor but not normal (but AF_N above 0)
* Orange: with contig in normal but not tumor (but AF_T above 0)

It's clear that there's a lot of variants with contigs assigned in both tumor and normal (thousands in each case), and including these would make the analysis quite noisy.

We have three options to deal with these variants with contigs in both samples...:
1. The safest option, keep all of them because we're scared about the inconsistency regarding AFs and we can't be sure that the variants with contigs in the normal aren't tumor contaminants
2. The most stringent option, discard all variants with a contig in the normal sample
3. Despite the inconsistencies of the AFs, apply the same AF filters as for the non-contig variants.

For a large number of these variants we have AFs, and they do follow the same pattern as those variants without contigs. In the scatterplot below both the tumor and the normal have assembled a contig for the variant (red bars in the barplot above):

<img width="1257" alt="image" src="https://user-images.githubusercontent.com/14308912/231155147-9e56d43f-13d9-4cde-9acb-f153d34b7b61.png">

The option I personally prefer is option 3, which means apply the AF-based filters to all variants, and when no AFs are available defer to the stringent filter of removing variants if there is a contig in the normal.

If we focus only on the Red box in the barplot above, the variants with contigs in both tumor and normal.
* If we apply the AF_filters mentioned earlier. We would filter away the variants in the Red box below (because higher than 25% TIN), and we would keep the variants in the small purple box (because lower than 25% TIN).
* For the remaining variants, the orange and blue, with a mix of missing AF values, we would filter these out because of the contig in the normal.
<img width="1215" alt="image" src="https://user-images.githubusercontent.com/14308912/231400340-63f63625-1564-4094-93b3-d53d1072b291.png">

For anyone who wants to suffer here's what the combination of different values and the decision of applying filters or not.
NormalAF_OK = Within TIN 25% threshold
NormalAF_HIGH = Above TIN 25% threshold
Keep these:
CTG_N_no:AF_N_no:CTG_T_no:AF_T_yes --> KEEP
CTG_N_no:AF_N_no:CTG_T_yes:AF_T_yes --> KEEP
CTG_N_no:AF_N_no:CTG_T_yes:AF_T_no --> KEEP
CTG_N_yes:AF_N_yes:CTG_T_yes:AF_T_yes:NormalAF_OK --> KEEP
CTG_N_no:AF_N_yes:CTG_T_no:AF_T_yes:NormalAF_OK --> KEEP
CTG_N_no:AF_N_yes:CTG_T_yes:AF_T_yes:NormalAF_OK --> KEEP
CTG_N_yes:AF_N_yes:CTG_T_no:AF_T_yes:NormalAF_OK --> KEEP
Filter out these:
CTG_N_yes:AF_N_yes:CTG_T_yes:AF_T_yes:NormalAF_HIGH --> filter out
CTG_N_no:AF_N_yes:CTG_T_no:AF_T_no --> filter out
CTG_N_yes:AF_N_yes:CTG_T_no:AF_T_no --> filter out
CTG_N_no:AF_N_yes:CTG_T_no:AF_T_yes:NormalAF_HIGH --> filter out
CTG_N_yes:AF_N_no:CTG_T_no:AF_T_no --> filter out
CTG_N_no:AF_N_yes:CTG_T_yes:AF_T_yes:NormalAF_HIGH --> filter out
CTG_N_yes:AF_N_no:CTG_T_yes:AF_T_no --> filter out
CTG_N_yes:AF_N_yes:CTG_T_no:AF_T_yes:NormalAF_HIGH --> filter out
CTG_N_yes:AF_N_no:CTG_T_yes:AF_T_yes --> filter out
CTG_N_yes:AF_N_no:CTG_T_no:AF_T_yes --> filter out
CTG_N_no:AF_N_yes:CTG_T_yes:AF_T_no --> filter out IF af_n > 0.20
CTG_N_yes:AF_N_yes:CTG_T_yes:AF_T_no --> filter out

And here are the barplots for these different groups of variants:
<img width="1228" alt="image" src="https://user-images.githubusercontent.com/14308912/231208623-9ad6609c-c273-440d-95c5-e96788ac4527.png">

Applying these filters would result in this:
<img width="1214" alt="image" src="https://user-images.githubusercontent.com/14308912/231208932-682a1bae-7f00-4343-aca5-3ece8ae46556.png">

Where this, previously seen scatterplot, shows the allele_frequencies:
<img width="1223" alt="image" src="https://user-images.githubusercontent.com/14308912/231140380-cc173f0f-859b-440d-b014-917dbe7caaed.png">

And these are only the PASS variants on the scatterplot, labeled according to their combination of properties:
<img width="1240" alt="image" src="https://user-images.githubusercontent.com/14308912/231210110-aec76c97-b04d-4b13-b498-8f8f7e5315d2.png">

And the corresponding barplot, showing that the vast majority of variants with PASS have no presence at all in the normal (as a result of these normal-samples having no TIN probably)
<img width="1208" alt="image" src="https://user-images.githubusercontent.com/14308912/231211147-38b6aa1e-4fae-4d3d-8f4e-397072bf30f0.png">

## BND variants

BND variant in TIDDIT has 2 calls, sometimes 1 of these are found in the normal, and the 2nd not. See below:

BND1: 2	218932779	SV_386_1
Tumor: ./.:2:63,70,84:1:1:0.0136986,0:72,71:94,97
Normal: ./.:.:.:.:.:.:.:.

BND2:2	218933210	SV_386_2
Tumor: ./.:2:63,70,84:1:1:0.0136986,0:72,71:94,97 
Normal: 0/1:2:27,26,28:0:0:0.0357143,0.0344828:20,27:31,36

If we applied the filters mentioned earlier to these variants we might cripple the annotation of some of them as both BNDs are required in VEP to correctly annotate the variants.

The best way to filter these variants would be to have a script that could look at the dual-BNDs as 1 (which is not possible with BCFtools since they are on different rows). This way we could filter out these variants only if BOTH BNDs are in the normal.

The faster option, which does not rely on anything more than bcftools would be to filter out only the BNDs that are ONLY in the normal. 

The faster option would mean that we would keep the variants in the green box, regardless if they are clearly germline variants. Which would add on average 600 variants extra as noise. 

<img width="1249" alt="image" src="https://user-images.githubusercontent.com/14308912/231212270-f3dea5ce-1fe6-4751-a1ed-39d697d34400.png">

### Extra figures:

Just for fun, number of SVtypes per case:

<img width="1149" alt="image" src="https://user-images.githubusercontent.com/14308912/231148813-86f8ae55-ae53-4ba2-90d4-739b83e33f82.png">

**Comment by mathiasbio** [comment]
Current state of the filters:

**normal_variant** set if:
AF_T_MAX == 0 and ctg_t == False

**high_normal_af** set if:
AF_N_MAX > 0.25

**high_normal_af_fraction** set if:
(AF_N_MAX / AF_T_MAX) > 0.25
    
**in_normal** set if:
ctg_n == True and AF_N_MAX == 0 and AF_T_MAX <= 0.25

<img width="1083" alt="image" src="https://user-images.githubusercontent.com/14308912/231975592-ce3c57d3-e5ba-48fb-b25a-f90818a2c1d5.png">

<img width="1097" alt="image" src="https://user-images.githubusercontent.com/14308912/231975695-93d8f97f-9be6-44d3-93bb-cdd0336b00ed.png">

**Comment by mathiasbio** [comment]
As I mentioned earlier, due to BND variants having 1 variants per BreakEnd, and the need for both to exist for proper annotation in VEP, and this combined with the observation that in some cases 1 of the BND variants are filtered as "in normal" and the other "PASS", I implemented this filter in a custom python script rather than using bcftools.

From tiddit_sv_tumor_normal rule:

**The first step annotates the variant with the normal-filters**
```
python {params.filter_svs} filter_tiddit_TN -v {params.vcfdir}/{params.tumor}_{params.normal}_merged_PASS.vcf \
-m filter > {params.vcfdir}/{params.tumor}_{params.normal}_merged_tn_filter.vcf ;
```

**The second step rescues BNDs if at least 1 out of the 2 breakends has PASS**
```
python {params.filter_svs} filter_tiddit_TN -v {params.vcfdir}/{params.tumor}_{params.normal}_merged_tn_filter.vcf \
-m rescue_bnds | bgzip -l 9 -c > {output.vcf}
```

Final filter assignment results:
<img width="726" alt="image" src="https://user-images.githubusercontent.com/14308912/233637313-26d0279b-a057-45c5-8a74-7360a192406d.png">

Scatterplot, as you can see there are a few variants with not allowed allele-frequencies because they're rescued due to its companion BND having filter PASS:
<img width="1082" alt="image" src="https://user-images.githubusercontent.com/14308912/233637395-b3dbcb39-01b7-4e07-8af8-97e58a7b1c15.png">

Perhaps more clearly observed here, keeping only the PASS variants:
<img width="1091" alt="image" src="https://user-images.githubusercontent.com/14308912/233637436-aad2c63a-75b8-4f1b-a48c-92a760975e77.png">

Very few BND-variants are rescued in this approach. PASS variants before and after:
* A: 3957 --> 3974 
* B: 3528  --> 3534
* C: 3872 --> 3887

**Comment by ivadym** [comment]
https://github.com/Clinical-Genomics/BALSAMIC/pull/1120 ✅
